src/bgw/job.c

```diff
@@ -0,0 +1,311 @@
+					void *data, MemoryContext mctx, LOCKMODE lockmode)
+{
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_bgw_job_stat_pkey_idx_job_id,
+				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(bgw_job_id));
+	bgw_job_scan_one(BGW_JOB_STAT_PKEY_IDX,
+					 scankey, 1, tuple_found, tuple_filter, data, mctx, lockmode);
+}
+
+static bool
+bgw_job_tuple_found(TupleInfo *ti, void *const data)
+{
+	BgwJob	  **job_pp = data;
+
+	*job_pp = bgw_job_from_tuple(ti->tuple, sizeof(BgwJob), ti->mctx);
+
+	/* return true since used with scan one */
+	return true;
+}
+
+BgwJob *
+bgw_job_find(int32 bgw_job_id, MemoryContext mctx)
+{
+	BgwJob	   *job_stat = NULL;
+
+	bgw_job_scan_job_id(bgw_job_id, bgw_job_tuple_found, NULL, &job_stat, mctx, AccessShareLock);
+
+	return job_stat;
+}
+
+#ifdef DEBUG
+#define TEST_JOB_FUNCTION "ts_bgw_job_execute_test"
+
+static bool
+call_test_dispatcher(BgwJob *job)
+{
+	/* the test dispatcher should throw an error if function not found */
+	PGFunction test_function = load_external_function(extension_get_so_name(), TEST_JOB_FUNCTION, false, NULL);
+
+	if (test_function == NULL)
+		elog(ERROR, "cannot find test execution function");
+
+	return DatumGetBool(DirectFunctionCall1(test_function, Int32GetDatum(job->fd.id)));
+}
+#endif
+
+bool
+bgw_job_execute(BgwJob *job)
+{
+	switch (job->bgw_type)
+	{
+		case JOB_TYPE_VERSION_CHECK:
+			/* TODO: fill in when merging with Amy's stuff  */
+			return false;
+		case _MAX_JOB_TYPE:
+			/* unkown job */
+#ifdef DEBUG
+			return call_test_dispatcher(job);
+#else
+			elog(ERROR, "unknown job type \"%s\"", NameStr(job->fd.job_type));
+#endif
+			break;
+	}
+	Assert(false);
+	/* Hack for some compilers */
+	return false;
+}
+
+bool
+bgw_job_has_timeout(BgwJob *job)
+{
+	Interval	zero_val = {.time = 0,};
+
+	return DatumGetBool(DirectFunctionCall2(interval_gt, IntervalPGetDatum(&job->fd.max_runtime), IntervalPGetDatum(&zero_val)));
+}
+
+/* Return the timestamp at which to kill the job due to a timeout */
+TimestampTz
+bgw_job_timeout_at(BgwJob *job, TimestampTz start_time)
+{
+	/* timestamptz plus interval */
+	return DatumGetTimestampTz(DirectFunctionCall2(
+												   timestamptz_pl_interval,
+												   TimestampTzGetDatum(start_time),
+												   IntervalPGetDatum(&job->fd.max_runtime)));
+}
+
+
+TS_FUNCTION_INFO_V1(ts_bgw_job_entrypoint);
+
+extern Datum
+ts_bgw_job_entrypoint(PG_FUNCTION_ARGS)
+{
+	Oid			db_oid = DatumGetObjectId(MyBgworkerEntry->bgw_main_arg);
+	int32		job_id = Int32GetDatum(DirectFunctionCall1(int4in, CStringGetDatum(MyBgworkerEntry->bgw_extra)));
+	BgwJob	   *job;
+	JobResult	res = JOB_FAILURE;
+
+	BackgroundWorkerBlockSignals();
+	/* Setup any signal handlers here */
+	BackgroundWorkerUnblockSignals();
+
+	elog(DEBUG1, "started background job %d", job_id);
+
+	BackgroundWorkerInitializeConnectionByOid(db_oid, InvalidOid);
+
+	/*
+	 * Start transactions early. Jobs are allowed to end and start transaction
```

Comment:
I think this comment needs to be updated given that we aren't doing the thing where the txn stays open for the duration of the job anymore.

Reply:
fixed